src/librbd/cache/SharedReadOnlyObjectDispatch.cc

```diff
@@ -0,0 +1,169 @@
+    ldout(cct, 5) << "child image: skipping" << dendl;
+    return;
+  }
+
+  ldout(cct, 5) << "parent image: setup SRO cache client" << dendl;
+
+  std::string controller_path = ((CephContext*)cct)->_conf.get_val<std::string>("immutable_object_cache_sock");
+  m_cache_client = new ceph::immutable_obj_cache::CacheClient(controller_path.c_str(), m_image_ctx->cct);
+  m_cache_client->run();
+
+  int ret = m_cache_client->connect();
+  if (ret < 0) {
+    ldout(cct, 5) << "SRO cache client fail to connect with local controller: "
+                  << "please start ceph-immutable-object-cache daemon"
+		  << dendl;
+  } else {
+    ldout(cct, 5) << "SRO cache client to register volume "
+                  << "name = " << m_image_ctx->id
+                  << " on ceph-immutable-object-cache daemon"
+                  << dendl;
+
+    auto ctx = new FunctionContext([this](bool reg) {
+      handle_register_client(reg);
+    });
+    ret = m_cache_client->register_client(ctx);
+
+    if (ret >= 0) {
+      // add ourself to the IO object dispatcher chain
+      m_image_ctx->io_object_dispatcher->register_object_dispatch(this);
+    }
+  }
+}
+
+template <typename I>
+bool SharedReadOnlyObjectDispatch<I>::read(
+    const std::string &oid, uint64_t object_no, uint64_t object_off,
+    uint64_t object_len, librados::snap_t snap_id, int op_flags,
+    const ZTracer::Trace &parent_trace, ceph::bufferlist* read_data,
+    io::ExtentMap* extent_map, int* object_dispatch_flags,
+    io::DispatchResult* dispatch_result, Context** on_finish,
+    Context* on_dispatched) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << "object_no=" << object_no << " " << object_off << "~"
+                 << object_len << dendl;
+
+  // if any session failed, reads will go to rados
+  if(!m_cache_client->is_session_work()) {
+    ldout(cct, 5) << "SRO cache client session failed " << dendl;
+    *dispatch_result = io::DISPATCH_RESULT_CONTINUE;
+    on_dispatched->complete(0);
+    return true;
+  }
+
+  auto ctx = new LambdaGenContext<std::function<void(ObjectCacheRequest*)>,
+      ObjectCacheRequest*>([this, snap_id, read_data, dispatch_result, on_dispatched,
+      oid, object_off, object_len](ObjectCacheRequest* ack) {
+
+    if (ack->type == RBDSC_READ_REPLY) {
+      std::string file_path = ((ObjectCacheReadReplyData*)ack)->cache_path;
+      ceph_assert(file_path != "");
+      handle_read_cache(file_path, object_off, object_len, read_data,
+                        dispatch_result, on_dispatched);
+    } else {
+      // go back to read rados
+      *dispatch_result = io::DISPATCH_RESULT_CONTINUE;
+      on_dispatched->complete(0);
+    }
+  });
+
+  if (m_cache_client && m_cache_client->is_session_work() && m_object_store) {
+
+    m_cache_client->lookup_object(m_image_ctx->data_ctx.get_namespace(),
+                                  m_image_ctx->data_ctx.get_id(),
+                                  (uint64_t)snap_id, oid, ctx);
+  }
+  return true;
```

Comment:
You cannot return ```true``` here if you didn't attempt to look-up the object

Reply:
fixed it. this is a bug. thanks for your suggestion.